modules/core/src/main/java/org/apache/ignite/internal/client/thin/ReliableChannel.java

```diff
@@ -473,6 +415,196 @@ public void addChannelFailListener(Runnable chFailLsnr) {
         chFailLsnrs.add(chFailLsnr);
     }
 
+    /** Should the channel initialization be stopped. */
+    private boolean stopInitCondition() {
+        return scheduledChannelsReinit.get() || closed;
+    }
+
+    /**
+     * Init channel holders to all nodes.
+     * @param force enable to replace existing channels with new holders.
+     */
+    private synchronized void initChannelHolders(boolean force) {
+        // enable parallel threads to schedule new init of channel holders
+        scheduledChannelsReinit.set(false);
+
+        if (!force && channels.get() != null)
+            return;
+
+        Set<InetSocketAddress> resolvedAddrs = parseAddresses(clientCfg.getAddresses());
```

Comment:
Somebody can configure addresses with DNS names, in k8s env it is possible with StatefulSet configuration. But node can change IP. So after topology change we should check resolve all addresses. Nevertheless, It's possible to cache result of resolution.

In case of static configuration holders are persisted. But it's true that cache help to win some by skip iteration steps.

Reply:
At least you can skip these steps if references to addresses are equals (in case of k8s you will get the new array each time, in case of static configuration - the same array each time)

Reply:
With StatefulSet it will be a static configuration, i.e. list of DNS names like "ignite-1, ignite-2, ignite-3". Also it's not prohibited to configure with DNS names in any other env (non-k8s). 

`TcpClientChannel` creates socket with `cfg.getAddress().getHostName()` and it invokes reverse lookup even in case we specify address with hostnames.

Kubernetes have troubles with reverse lookup, as pods aren't aimed to be reverse resolved at all. There are some work in that direction, but the issue isn't resolved yet: https://github.com/coredns/coredns/issues/3888

So, I think we can't rely on this reverse lookup. So it could be case when DNS name isn't changed but the channel is broken and don't have changes to be fixed. So re-resolution is required.

I will check that resolved address isn't changed. And so skip next steps.

Reply:
To solve this problem we can store address as unresolved (InetSocketAddress#createUnresolved).
In your implementation initChannelHolders will be invoked in response reading thread on each topology change. I think address resolving in this thread can take some time and should be avoided.

Reply:
Fixed.